cloudinit/config/cc_ca_certs.py

```diff
@@ -100,95 +121,133 @@ def update_ca_certs(distro_cfg):
-    Update Certificate config file to add the file path managed cloud-init
+    For every entry in the CA_CERT_CONFIG file prefix the entry with a "!"
+    in order to disable it.
 
     @param distro_cfg: A hash providing _distro_ca_certs_configs function.
     """
     if distro_cfg["ca_cert_config"] is None:
         return
-    if os.stat(distro_cfg["ca_cert_config"]).st_size == 0:
-        # If the CA_CERT_CONFIG file is empty (i.e. all existing
-        # CA certs have been deleted) then simply output a single
-        # line with the cloud-init cert filename.
-        out = "%s\n" % distro_cfg["ca_cert_filename"]
-    else:
-        # Append cert filename to CA_CERT_CONFIG file.
-        # We have to strip the content because blank lines in the file
-        # causes subsequent entries to be ignored. (LP: #1077020)
+    if os.stat(distro_cfg["ca_cert_config"]).st_size != 0:
         orig = util.load_file(distro_cfg["ca_cert_config"])
-        cr_cont = "\n".join(
-            [
-                line
-                for line in orig.splitlines()
-                if line != distro_cfg["ca_cert_filename"]
-            ]
-        )
-        out = "%s\n%s\n" % (cr_cont.rstrip(), distro_cfg["ca_cert_filename"])
+        out = ""
+        for line in orig.splitlines():
+            if line.startswith("#") or line == "":
+                out += line + "\n"
+            else:
+                out += "!" + line + "\n"
     util.write_file(distro_cfg["ca_cert_config"], out, omode="wb")
 
 
-def remove_default_ca_certs(distro_name, distro_cfg):
+def remove_default_ca_certs(distro_cfg):
     """
-    Removes all default trusted CA certificates from the system. To actually
-    apply the change you must also call L{update_ca_certs}.
+    Removes all default trusted CA certificates from the system.
 
-    @param distro_name: String providing the distro class name.
     @param distro_cfg: A hash providing _distro_ca_certs_configs function.
     """
+    if distro_cfg["ca_cert_path"] is None:
+        return
+
+    LOG.debug("Deleting system CA certificates")
     util.delete_dir_contents(distro_cfg["ca_cert_path"])
-    util.delete_dir_contents(distro_cfg["ca_cert_system_path"])
-    util.write_file(distro_cfg["ca_cert_config"], "", mode=0o644)
 
-    if distro_name in ["debian", "ubuntu"]:
-        debconf_sel = (
-            "ca-certificates ca-certificates/trust_new_crts " + "select no"
-        )
-        subp.subp(("debconf-set-selections", "-"), debconf_sel)
+
+def move_freebsd_ca_certs_to_blacklist_dir(distro_cfg):
+    """
+    For every file in the CA_CERT_PATH directory move it to the
+    CA_CERT_BLACKLIST_DIR in order to disable it.
+
+    @param distro_cfg: A hash providing _distro_ca_certs_configs function.
+    """
+    if (
+        distro_cfg["ca_cert_path"] is None
+        or distro_cfg["ca_cert_blacklist_path"] is None
+    ):
+        return
+    LOG.debug("Moving system CA certificates to blacklist directory")
+    shutil.move(
```

Comment:
this is, unfortunately, not idempotent:

```python-traceback
2023-02-01 00:10:59,257 - cc_ca_certs.py[DEBUG]: Moving system CA certificates to blacklist directory
2023-02-01 00:10:59,260 - handlers.py[DEBUG]: finish: init-network/config-ca-certs: FAIL: running config-ca-certs with frequency once-per-instance
2023-02-01 00:10:59,260 - util.py[WARNING]: Running module ca-certs (<module 'cloudinit.config.cc_ca_certs' from '/usr/local/lib/python3.9/site-packages/cloudinit/config/cc_ca_certs.py'>) failed
2023-02-01 00:10:59,261 - util.py[DEBUG]: Running module ca-certs (<module 'cloudinit.config.cc_ca_certs' from '/usr/local/lib/python3.9/site-packages/cloudinit/config/cc_ca_certs.py'>) failed
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/shutil.py", line 825, in move
    os.rename(src, real_dst)
FileNotFoundError: [Errno 2] No such file or directory: '/usr/share/certs/trusted/' -> '/usr/share/certs/blacklisted/trusted'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/cloudinit/config/modules.py", line 246, in _run_modules
    ran, _r = cc.run(
  File "/usr/local/lib/python3.9/site-packages/cloudinit/cloud.py", line 67, in run
    return self._runners.run(name, functor, args, freq, clear_on_fail)
  File "/usr/local/lib/python3.9/site-packages/cloudinit/helpers.py", line 185, in run
    results = functor(*args)
  File "/usr/local/lib/python3.9/site-packages/cloudinit/config/cc_ca_certs.py", line 269, in handle
    disable_default_ca_certs(cloud.distro.name, distro_cfg)
  File "/usr/local/lib/python3.9/site-packages/cloudinit/config/cc_ca_certs.py", line 161, in disable_default_ca_certs
    move_freebsd_ca_certs_to_blacklist_dir(distro_cfg)
  File "/usr/local/lib/python3.9/site-packages/cloudinit/config/cc_ca_certs.py", line 220, in move_freebsd_ca_certs_to_blacklist_dir
    shutil.move(
  File "/usr/local/lib/python3.9/shutil.py", line 845, in move
    copy_function(src, real_dst)
  File "/usr/local/lib/python3.9/shutil.py", line 444, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/local/lib/python3.9/shutil.py", line 264, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/usr/share/certs/trusted/'
```

let's pre-condition this `shutil.move()` on `distro_cfg["ca_cert_path"]` actually existing.

Reply:
at the very least, we need to catch an exceptions here.

Reply:
Removed FreeBSD related changes.